### PR TITLE
dev/core#2682 Entity batch declare option group for pseudoconstant

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
dev/core#2682 Entity batch declare option group for pseudoconstant

Before
----------------------------------------
Values for entity_table not discoverable

After
----------------------------------------
Discoverable

Technical Details
----------------------------------------
refer https://lab.civicrm.org/dev/core/-/issues/2682

Comments
----------------------------------------
Part of apiv4 entity parity

### CODE REVIEW

File: xml/templates/civicrm_data.tpl [hunk 1]
```diff
@@ -284,6 +285,8 @@ SELECT @option_group_id_ps    := max(id) from civicrm_option_group where name =
 SELECT @option_group_id_crs    := max(id) from civicrm_option_group where name = 'contribution_recur_status';
 SELECT @option_group_id_env    := max(id) from civicrm_option_group where name = 'environment';
 SELECT @option_group_id_default_assignee := max(id) from civicrm_option_group where name = 'activity_default_assignee';
+SELECT @option_group_id_default_assignee := max(id) from civicrm_option_group where name = 'activity_default_assignee';
```
Comment (seamuslee001) [inline xml/templates/civicrm_data.tpl:288]: @eileenmcnaughton this looks like it is duplicated in this line

Comment (eileenmcnaughton) [reply xml/templates/civicrm_data.tpl]: thanks @seamuslee001 I removed it - but I think there is something else wrong in here. So far I've tested the upgrade locally but `civibuild create` - so it's probably something obvious

File: CRM/Upgrade/Incremental/Base.php [hunk 2]
```diff
@@ -220,6 +220,30 @@ public static function addColumn($ctx, $table, $column, $properties, $localizabl
...
+   * @param array $options
+   *
+   * @return bool
+   */
+  public static function addOptionGroup(CRM_Queue_TaskContext $ctx, $params, $options): bool {
+    $defaults = [
+      'is_reserved' => 1,
+      'is_active' => 1,
+      'is_locked' => 1,
+    ];
```
Comment (colemanw) [inline CRM/Upgrade/Incremental/Base.php:237]: These defaults seem more like "gotchas" for those not expecting them to be auto-applied :/
I think upgrade functions would be easier to grok if every attribute was explicitly passed in.

Comment (eileenmcnaughton) [reply CRM/Upgrade/Incremental/Base.php]: @colemanw yeah - I left is_active there but moved the other 2 out - seems like you wouldn't create them inactive & if you did that would be pretty intentional

**Comment by colemanw** [comment]
>  I left is_active there but moved the other 2 out - seems like you wouldn't create them inactive & if you did that would be pretty intentional

Ok. I think that's already the API default.

File: xml/templates/civicrm_data.tpl [hunk 3]
```diff
@@ -1054,7 +1056,12 @@ VALUES
...
 (@option_group_id_default_assignee, '{ts escape="sql"}Specific contact{/ts}',               '3',     'SPECIFIC_CONTACT',        NULL,       0,         0,           1,         NULL,          0,             0,             1,           NULL,            NULL,           NULL),
-(@option_group_id_default_assignee, '{ts escape="sql"}User creating the case{/ts}',          '4',     'USER_CREATING_THE_CASE',  NULL,       0,         0,           1,         NULL,          0,             0,             1,           NULL,            NULL,           NULL);
+(@option_group_id_default_assignee, '{ts escape="sql"}User creating the case{/ts}',          '4',     'USER_CREATING_THE_CASE',  NULL,       0,         0,           1,         NULL,          0,             0,             1,           NULL,            NULL,           NULL),
+
+-- Entity Batch options
+-  (`option_group_id`,             `label`,                                      `value`,                   `name`,                    `grouping`, `filter`, `is_default`, `weight`, `description`, `is_optgroup`, `is_reserved`, `is_active`, `component_id`, `visibility_id`, `icon`)
```
Comment (seamuslee001) [inline xml/templates/civicrm_data.tpl:1062]: @eileenmcnaughton this line needs a 2nd `-` at the start of it

Comment (eileenmcnaughton) [reply xml/templates/civicrm_data.tpl]: @seamuslee001 ah thanks - I knew there was a bug in there - you saved me tracking it down

**Comment by eileenmcnaughton** [comment]
progress - it's just failing apiv4 syntax conformance now

**Comment by eileenmcnaughton** [comment]
@seamuslee001 do you have any idea how to fix this - this test won't run at all for me locally (with or without this patch)

**Comment by seamuslee001** [comment]
@eileenmcnaughton I think I got it working now, I had to add in a default to the entity_table in v4 as the GET request it was failing on was generating sql like 

```
Aug 30 08:30:57  [debug] $Query = SELECT `a`.`id` AS `id`, `a`.`entity_table` AS `entity_table`, `a`.`entity_id` AS `entity_id`, `a`.`batch_id` AS `batch_id`
FROM civicrm_entity_batch a
WHERE ((`a`.`id` IS NULL OR (`a`.`id` IN (SELECT id FROM `civicrm_entity_batch` WHERE ((entity_table = 'civicrm_financial_trxn')))))) AND (`a`.`id` = "1")
```

but in the civicrm_entity_batch the entity_table was NULL

Can you check what I have done and if your ok with It I think this is now in a mergable state

**Comment by eileenmcnaughton** [comment]
@seamuslee001 I think that's OK

**Comment by colemanw** [comment]
Nice work @eileenmcnaughton & @seamuslee001 
I like the new upgrader function.

**Pull request merged**